File: raytrace/src/rays.rs
use crate::tuple::Tuple;
use crate::matrix::Matrix;

#[derive(Debug,Clone,Copy)]
pub struct Ray {
    pub origin: Tuple,
    pub direction: Tuple,
}

impl Ray {
    pub fn position(&self, distance: f64) -> Tuple {
        self.origin + (*&self.direction * distance)
    }

    pub fn transform(&self, t: Matrix) -> Self {
        ray( t.multup( &self.origin ),
             t.multup( &self.direction ))
    }
}

pub fn ray(origin: Tuple, direction: Tuple) -> Ray {
    Ray { origin, direction }
}

#[cfg(test)]
mod tests {
    use crate::rays::ray;
    use crate::tuple::{point, vector};
    use crate::transform::{translation, scaling};

    #[test]
    fn new_creates_rays(){
        let o = point( 1.0, 2.0, 3.0);
        let d = vector( 4.0, 5.0, 6.0);

        let r = ray(o, d);
        assert!( r.origin.equals(point( 1.0, 2.0, 3.0))) ;
        assert!( r.direction.equals(vector( 4.0, 5.0, 6.0))) ;
    }

    #[test]
    fn computing_point_from_distance(){
        let o = point( 2.0, 3.0, 4.0);
        let d = vector( 1.0, 0.0, 0.0);

        let r = ray(o, d);
        assert!( r.position(0.0).equals(point( 2.0, 3.0, 4.0))) ;
        assert!( r.position(1.0).equals(point( 3.0, 3.0, 4.0))) ;
        assert!( r.position(-1.0).equals(point( 1.0, 3.0, 4.0))) ;
        assert!( r.position(2.5).equals(point( 4.5, 3.0, 4.0))) ;

    }

    #[test]
    fn translating_a_ray(){
        let o = point( 1.0, 2.0, 3.0);
        let d = vector( 0.0, 1.0, 0.0);
        let r = ray(o, d);

        let m = translation(3.0, 4.0, 5.0);
        let r2 = r.transform(m);

        assert!( r2.origin.equals(point( 4.0, 6.0, 8.0))) ;
        assert!( r2.direction.equals(vector( 0.0, 1.0, 0.0))) ;
    }

    #[test]
    fn scaling_a_ray(){
        let o = point( 1.0, 2.0, 3.0);
        let d = vector( 0.0, 1.0, 0.0);
        let r = ray(o, d);

        let m = scaling(2.0, 3.0, 4.0);
        let r2 = r.transform(m);

        assert!( r2.origin.equals(point( 2.0, 6.0, 12.0))) ;
        assert!( r2.direction.equals(vector( 0.0, 3.0, 0.0))) ;
    }
}

File: python/rays.py
# to run tests: python -m unittest -v rays

import unittest
import tuples
import transformations

class Ray:
    def __init__(self, origin, direction):
        self.origin = origin
        self.direction = direction

    def position(self, t):
        return self.origin + self.direction * t

    def transform(self, matrix):
        origin = matrix * self.origin
        direction = matrix * self.direction
        return ray(origin, direction)

def ray(origin, direction):
    return Ray(origin, direction)

class RayTestCase(unittest.TestCase):
    def test_creating_and_querying_a_ray(self):
        origin = tuples.point(1, 2, 3)
        direction = tuples.vector(4, 5, 6)

        r = ray(origin, direction)

        self.assertEqual(r.origin, origin)
        self.assertEqual(r.direction, direction)

    def test_computing_a_point_from_a_distance(self):
        r = ray(tuples.point(2, 3, 4), tuples.vector(1, 0, 0))

        self.assertEqual(r.position(0), tuples.point(2, 3, 4))
        self.assertEqual(r.position(1), tuples.point(3, 3, 4))
        self.assertEqual(r.position(-1), tuples.point(1, 3, 4))
        self.assertEqual(r.position(2.5), tuples.point(4.5, 3, 4))

    def test_translating_a_ray(self):
        r = ray(tuples.point(1, 2, 3), tuples.vector(0, 1, 0))
        m = transformations.translation(3, 4, 5)

        r2 = r.transform(m)

        self.assertEqual(r2.origin, tuples.point(4, 6, 8))
        self.assertEqual(r2.direction, tuples.vector(0, 1, 0))
        
    def test_scaling_a_ray(self):
        r = ray(tuples.point(1, 2, 3), tuples.vector(0, 1, 0))
        m = transformations.scaling(2, 3, 4)

        r2 = r.transform(m)

        self.assertEqual(r2.origin, tuples.point(2, 6, 12))
        self.assertEqual(r2.direction, tuples.vector(0, 3, 0))
        
# ---------------------------------------------------------------------------
if __name__ == '__main__':
    unittest.main()
